File: src/main/java/svnserver/repository/git/GitEntryImpl.java
/**
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.repository.git;

import org.eclipse.jgit.lib.FileMode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.tmatesoft.svn.core.SVNException;
import svnserver.StringHelper;
import svnserver.repository.VcsFile;
import svnserver.repository.git.prop.GitProperty;

import java.io.IOException;

/**
 * Simple GitEntry implementation.
 *
 * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
 */
public class GitEntryImpl implements GitEntry {
  @NotNull
  private final GitProperty[] props;
  @NotNull
  private final String name;
  @NotNull
  private final String parentPath;

  // Cache
  @Nullable
  private String fullPathCache;

  public GitEntryImpl(@NotNull GitProperty[] parentProps, @NotNull String parentPath, @NotNull GitProperty[] props, @NotNull String name, @NotNull FileMode fileMode) {
    this.parentPath = parentPath;
    this.name = name;
    this.props = GitProperty.joinProperties(parentProps, name, fileMode, props);
  }

  @NotNull
  @Override
  public GitProperty[] getRawProperties() {
    return props;
  }

  @NotNull
  @Override
  public String getFileName() {
    return name;
  }

  @NotNull
  @Override
  public String getFullPath() {
    if (fullPathCache == null) {
      fullPathCache = StringHelper.joinPath(parentPath, getFileName());
    }
    return fullPathCache;
  }

  @Nullable
  @Override
  public VcsFile getEntry(@NotNull String name) throws IOException, SVNException {
    return null;
  }

  @NotNull
  @Override
  public GitEntry createChild(@NotNull String name, boolean isDir) {
    return new GitEntryImpl(props, getFullPath(), GitProperty.emptyArray, name, isDir ? FileMode.TREE : FileMode.REGULAR_FILE);
  }
}

File: src/main/kotlin/svnserver/repository/git/GitEntryImpl.kt
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.repository.git

import org.eclipse.jgit.lib.FileMode
import svnserver.StringHelper
import svnserver.repository.git.prop.GitProperty
import java.io.IOException

/**
 * Simple GitEntry implementation.
 *
 * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
 */
internal open class GitEntryImpl(parentProps: Array<GitProperty>, parentPath: String, props: Array<GitProperty>, final override val fileName: String, fileMode: FileMode, stringInterner: (String) -> String) : GitEntry {
    final override val fullPath: String
    final override val rawProperties: Array<GitProperty> = GitProperty.joinProperties(parentProps, fileName, fileMode, props)

    init {
        fullPath = stringInterner(StringHelper.joinPath(parentPath, fileName))
    }

    override fun createChild(name: String, isDir: Boolean, stringInterner: (String) -> String): GitEntry {
        return GitEntryImpl(rawProperties, fullPath, GitProperty.emptyArray, name, if (isDir) FileMode.TREE else FileMode.REGULAR_FILE, stringInterner)
    }

    @Throws(IOException::class)
    override fun getEntry(name: String, stringInterner: (String) -> String): GitFile? {
        return null
    }
}
